### PR TITLE
Hclang 41 initial ebnf

### PR DESCRIPTION
**Pull request by Lambcreek**
EBNF and CFGs for our language

### Changelog
 Added two .txt files, one for EBNF, and one for CFGs. They should be "equivalent".

### Issues solved 
 - HCLANG-41

Please be very thorough when reviewing this. The CFG is very important for the project, and I most likely forgot some important things while I made it.

### CODE REVIEW

File: misc/CFGs/CFG.txt [hunk 1]
```diff
@@ -0,0 +1,89 @@
+**************************************************************
+CFG
+**************************************************************
+
+Program --> Start $
+
+Start --> FunctionDcl Start
```
Comment (C0DK) [inline misc/CFGs/CFG.txt]: evt kald den noget andet end "start".. evt "command" eller "statement"

File: misc/CFGs/CFG.txt [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+BracketR --> ]
+
+Type --> number
+       | text
+       | tuple
+       | list
+       | dict
+       | char
+       | none
+       | lambda
```
Comment (C0DK) [inline misc/CFGs/CFG.txt]: kalder vi den ikke "func" og ikke "lambda" :dancer:

File: misc/CFGs/CFG.txt [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+ParenR   --> )
+BracketL --> [
+BracketR --> ]
+
+Type --> number
+       | text
+       | tuple
+       | list
+       | dict
+       | char
```
Comment (C0DK) [inline misc/CFGs/CFG.txt]: skal vi have char? jeg tænkte at en char bare er en streng på 1 "character" uden at char faktisk er accessable fra en bruger synsvinkel. For simplicitet :P

File: misc/CFGs/CFG.txt (hunk 1 above)
Comment (C0DK) [reply misc/CFGs/CFG.txt]: bør denne ikke bare være statements?

File: misc/CFGs/CFG.txt [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+Expression --> FunctionCall
+             | Value
+
+Value --> Literal
+        | identifier
+
+Literal --> literalNumber
+          | literalText
+          | literalChar
```
Comment (C0DK) [inline misc/CFGs/CFG.txt]: igen skal vi bruge char?

File: misc/CFGs/CFG.txt [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+               | FunCallParameter identifier FunCallParamList
+
+FunCallParamList --> FunCallParameter FunCallParameterList
+                   | ParenL FunCallParameter FunCallParameterList ParenR
+
+FunCallParameter --> Value
+                   | ::identifier
+                   | LambdaExpression
+
+LambdaExpression --> ParenL FunDclParamList ParenR -> Type CurlyL Statements CurlyR
```
Comment (C0DK) [inline misc/CFGs/CFG.txt]: mangler der ikke mulighed for fx bare at lave:
[1,2,3,4] where { value == 5 } fx. altså uden "ParenL ParenR -> Type ". den type skal selvføgelig dog ikke være mulig i FunctionDcl.. :thinking:

File: misc/CFGs/CFG.txt (hunk 1 above)
Comment (C0DK) [reply misc/CFGs/CFG.txt]: dog uden "return" i guess :laughing:

File: misc/CFGs/HCLANG_EBNF.txt [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+ParenR   --> )
+BracketL --> [
+BracketR --> ]
+
+Type --> number
+       | text
+       | tuple
+       | list
+       | dict
+       | char
```
Comment (C0DK) [inline misc/CFGs/HCLANG_EBNF.txt]: skal vi have char? [igen]

File: misc/CFGs/HCLANG_EBNF.txt [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+Expression --> FunctionCall
+             | Value
+
+Value --> Literal
+        | identifier
+
+Literal --> literalNumber
+          | literalText
+          | literalChar
```
Comment (C0DK) [inline misc/CFGs/HCLANG_EBNF.txt]: char? [igen]

File: misc/CFGs/HCLANG_EBNF.txt [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+something:    1 occurrence of something
+[something]:  something 0 or 1 times
+{something}:  something 0 or more times
+{something}+: something 1 or more times
+
+***********************************************************************************
+
+Program --> Start $
+
+Start --> {FunctionDcl | VarDcl}
```
Comment (C0DK) [inline misc/CFGs/HCLANG_EBNF.txt]: her går min "start" kommentar fra CFG igen

File: misc/CFGs/HCLANG_EBNF.txt [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+               | FunCallParameter identifier FunCallParamList
+
+FunCallParamList --> {FunCallParameter}+
+                   | ParenL {FunCallParameter}+ ParenR
+
+FunCallParameter --> Value
+                   | ::identifier
+                   | LambdaExpression
+
+LambdaExpression --> ParenL [FunDclParamList] ParenR -> Type CurlyL Statements CurlyR
```
Comment (C0DK) [inline misc/CFGs/HCLANG_EBNF.txt]: her mangler også muligheden, når lambda blir parsed til en funktion, for at lave implicit types

File: misc/CFGs/HCLANG_EBNF.txt [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+           | bracketR
+           | parenL
+           | parenR
+           | curlyL
+           | curlyR
+           | assign
+           | functionPassOp
+           | arrow
+
+Type --> number
```
Comment (technight) [inline misc/CFGs/HCLANG_EBNF.txt]: Der mangler func type

File: misc/CFGs/HCLANG_EBNF.txt [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+Statement --> VarDcl linebreak
+            | VarAssign linebreak
+            | Expression linebreak
+            | ReturnStatement linebreak
+
+VarDcl --> Type identifier [assign Expression]
+
+VarAssign --> identifier assign Expression
+
+FunctionDcl --> func identifier assign parenL [FunDclParamList] parenR arrow Type curlyL Statements curlyR
```
Comment (technight) [inline misc/CFGs/HCLANG_EBNF.txt]: Er stadig lidt forvirret over, at der er funcDcl når det nu bare er det samme som med var?

File: misc/CFGs/HCLANG_EBNF.txt [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+Type --> number
+       | text
+       | tuple
+       | list
+       | dict
+       | none
+       | var
+
+Expression --> FunctionCall
+             | Value
```
Comment (technight) [inline misc/CFGs/HCLANG_EBNF.txt]: Skal Expression ikke også kunne holde lambda expression?

File: misc/CFGs/HCLANG_EBNF.txt (hunk 10 above)
Comment (technight) [reply misc/CFGs/HCLANG_EBNF.txt]: Og måske bool også

File: misc/CFGs/CFG.txt [hunk 13]
```diff
@@ -0,0 +1,87 @@
+**************************************************************
+CFG
+**************************************************************
+
+Program --> Commands $
+
+Commands --> FunctionDcl Commands
```
Comment (C0DK) [inline misc/CFGs/CFG.txt]: change this into "commands --> command commands" 
and then have command be defined as:
"command --> declaration | assigment | expression | EPSILON

File: misc/CFGs/CFG.txt [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+         | EPSILON
+
+Operator --> bracketL
+           | bracketR
+           | parenL
+           | parenR
+           | curlyL
+           | curlyR
+           | assign
+           | functionPassOp
```
Comment (C0DK) [inline misc/CFGs/CFG.txt]: This operator is not used. was obsolete

File: misc/CFGs/CFG.txt [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+            | VarAssign
+            | Expression
+            | ReturnStatement
+
+VarDcl --> Type identifier
+         | Type VarAssign
+
+VarAssign --> identifier assign Expression
+
+FunctionDcl --> FunDclStart identifier assign LambdaParams curlyL Statements curlyR
```
Comment (C0DK) [inline misc/CFGs/CFG.txt]: func dcl is another kind of varassign. not two different things.

File: misc/CFGs/CFG.txt [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+
+FunctionCall --> identifier
+               | FunCallParameter identifier FunCallParamList
+
+FunCallParamList --> FunCallParameter FunCallParameterList
+                   | parenL FunCallParameter FunCallParameterList parenR
+                   | EPSILON
+
+FunCallParameter --> Value
+                   | functionPassOp identifier
```
Comment (C0DK) [inline misc/CFGs/CFG.txt]: A function can be parsed by just using an identifer - functionPassOp is not needed.

File: misc/CFGs/CFG.txt [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+         | Type VarAssign
+
+VarAssign --> identifier assign Expression
+
+FunctionDcl --> FunDclStart identifier assign LambdaParams curlyL Statements curlyR
+
+FunDclStart --> func
+              | var
+
+FunDclParamList --> FunDclParameter FunDclSecondaryParameter
```
Comment (C0DK) [inline misc/CFGs/CFG.txt]: maybe place this near what it's relevant for. it's a bit confusing currently.

File: misc/CFGs/HCLANG_EBNF.txt [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+Commands --> {FunctionDcl | VarDcl | VarAssign | Expression}
+
+Operator --> bracketL
+           | bracketR
+           | parenL
+           | parenR
+           | curlyL
+           | curlyR
+           | assign
+           | functionPassOp
```
Comment (C0DK) [inline misc/CFGs/HCLANG_EBNF.txt]: not relevant

File: misc/CFGs/HCLANG_EBNF.txt [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+something:    1 occurrence of something
+[something]:  something 0 or 1 times
+{something}:  something 0 or more times
+{something}+: something 1 or more times
+
+***********************************************************************************
+
+Program --> Commands $
+
+Commands --> {FunctionDcl | VarDcl | VarAssign | Expression}
```
Comment (C0DK) [inline misc/CFGs/HCLANG_EBNF.txt]: functionDcl == vardcl

File: misc/CFGs/HCLANG_EBNF.txt [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+
+FunctionCall --> identifier
+               | FunCallParameter identifier
+               | FunCallParameter identifier FunCallParamList
+
+FunCallParamList --> {FunCallParameter}
+                   | parenL {FunCallParameter} parenR
+
+FunCallParameter --> Value
+                   | functionPassOp identifier
```
Comment (C0DK) [inline misc/CFGs/HCLANG_EBNF.txt]: see CFG comment

**Pull request merged**